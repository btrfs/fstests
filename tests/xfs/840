#! /bin/bash
# SPDX-License-Identifier: GPL-2.0
# Copyright (c) 2025 Oracle.  All Rights Reserved.
#
# FS QA Test No. 840
#
# hardware large atomic writes error inject test
#
. ./common/preamble
_begin_fstest auto rw quick atomicwrites

. ./common/filter
. ./common/inject
. ./common/atomicwrites
. ./common/reflink

_require_scratch_write_atomic
_require_scratch_write_atomic_multi_fsblock
_require_xfs_io_command "statx" "-r"
_require_xfs_io_command pwrite -A
_require_xfs_io_error_injection "bmap_finish_one"
_require_cp_reflink
_require_scratch_reflink

_scratch_mkfs >> $seqres.full 2>&1
_scratch_mount

echo "Create files"
file1=$SCRATCH_MNT/file1
touch $file1

max_awu=$(_get_atomic_write_unit_max $file1)
test $max_awu -ge 4096 || _notrun "cannot perform 4k atomic writes"

file2=$SCRATCH_MNT/file2
_pwrite_byte 0x66 0 64k $SCRATCH_MNT/file1 >> $seqres.full
cp --reflink=always $file1 $file2

echo "Check files"
md5sum $SCRATCH_MNT/file1 | _filter_scratch
md5sum $SCRATCH_MNT/file2 | _filter_scratch

echo "Inject error"
_scratch_inject_error "bmap_finish_one"

echo "Atomic write to a reflinked file"
$XFS_IO_PROG -dc "pwrite -A -D -V1 -S 0x67 0 4096" $file1

echo "FS should be shut down, touch will fail"
touch $SCRATCH_MNT/badfs 2>&1 | _filter_scratch

echo "Remount to replay log"
_scratch_remount_dump_log >> $seqres.full

echo "Check files"
md5sum $SCRATCH_MNT/file1 | _filter_scratch
md5sum $SCRATCH_MNT/file2 | _filter_scratch

echo "FS should be online, touch should succeed"
touch $SCRATCH_MNT/goodfs 2>&1 | _filter_scratch

# success, all done
status=0
exit
