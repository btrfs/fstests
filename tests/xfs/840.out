QA output created by 840
Create files
Check files
77e3a730e3c75274c9ce310d7e39f938  SCRATCH_MNT/file1
77e3a730e3c75274c9ce310d7e39f938  SCRATCH_MNT/file2
Inject error
Atomic write to a reflinked file
pwrite: Input/output error
FS should be shut down, touch will fail
touch: cannot touch 'SCRATCH_MNT/badfs': Input/output error
Remount to replay log
Check files
0df1f61ed02a7e9bee2b8b7665066ddc  SCRATCH_MNT/file1
77e3a730e3c75274c9ce310d7e39f938  SCRATCH_MNT/file2
FS should be online, touch should succeed
